README.md
# Icontrol

Gem used to wrap iControl API REST calls to an F5. The intent of this client is to match the iControl API methods and pass the response without modification. It does not contain business logic to scrub a response or prevent certain methods. This logic is to be in your orchestration layer, not in the client.

This has been tested against 11.5.4, but should be compatible with future versions. If this gem isn't compatible with a newer version of BIG-IP, or doesn't have the functionality required to access a certain method, feel free to follow the [guidelines for contributing](CONTRIBUTING.md) and provide an enhancement. When doing so, please follow semantic versioning when considering changes.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'icontrol_rest'
```

And then execute:

```
$ bundle install
```

Or install it yourself:

```
$ gem install icontrol_rest
```


## Usage

```ruby
require 'icontrol_rest'

api = IcontrolRest::Client.new(host: '1.2.3.4', user: 'user', pass: 'pass', verify_cert: false)
# => <IcontrolRest::Client:0x007fb953ab7750 @options={ ... }>

api.get_sys_dns
# => {"kind"=>"tm:sys:dns:dnsstate",
      "selfLink"=>"https://localhost/mgmt/tm/sys/dns?ver=11.5.4",
      "description"=>"configured-by-dhcp",
      "nameServers"=>["1.2.3.72", "1.2.3.73"],
      "search"=>["domain.com"]}

# Or you can manually do it:
api.get('/mgmt/tm/sys/dns')
# => {"kind"=>"tm:sys:dns:dnsstate",
      "selfLink"=>"https://localhost/mgmt/tm/sys/dns?ver=11.5.4",
      "description"=>"configured-by-dhcp",
      "nameServers"=>["1.2.3.72", "1.2.3.73"],
      "search"=>["domain.com"]}

# post/put/patch will allow you to specify a body and header with your request
api.post_transaction(body: {})
# => {"transId"=>1526674837169915,
      "state"=>"STARTED",
      "timeoutSeconds"=>120,
      "asyncExecution"=>false,
      "validateOnly"=>false,
      "executionTimeout"=>300,
      "executionTime"=>0,
      "failureReason"=>"",
      "kind"=>"tm:transactionstate",
      "selfLink"=>"https://localhost/mgmt/tm/transaction/1526674837169915?ver=12.1.2"}

api.put_sys_dns(body: { nameServers: ['1.2.3.69'] }, headers: { 'X-F5-REST-Coordination-Id' => '1526674837169915' })
# => {"method": "PUT",
      "uri": "https://localhost/mgmt/tm/sys/dns",
      "body": {
          "nameServers": ["1.2.3.69"]
      },
      "evalOrder": 1,
      "commandId": 1,
      "kind": "tm:transaction:commandsstate",
      "selfLink": "https://localhost/mgmt/tm/transaction/1526674837169915/commands/1?ver=12.1.2"
```


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `bundle exec rspec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.


## Contributing

Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.


CHANGELOG.md
# CHANGELOG

## 3.0.1

* Uplift for Ruby 2.7 support.
* Add double splat to method_missing so that explicit hash is passed.

## 3.0.0

* Require 'body' and 'headers' keyword arguments when they should be provided to any request type.
* Add support for custom headers.
* Update Logging framework.

## 2.3.0

* Adding patch method to client.

## 2.1.2

* Making retries an optional parameter.
* Adding automatic retries when the API returns an XML response.
* Adding an optional parameter to allow a user to sleep after each call.
* Adding bare bones logging to the client.
* Adding option to pass timeout.


spec/client_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe IcontrolRest::Client do
  let(:api) { IcontrolRest::Client.new(host: '1.2.3.4', user: 'user', pass: 'pass', verify_cert: false) }
  let(:bad_response) { double('Object', code: 500, :[] => '') }

  describe '#get/post/put/delete' do
    context 'when an empty route chain is passed' do
      it 'raises StandardError' do
        expect { api.get_ }.to raise_error StandardError
      end
    end

    context 'when anything but a 200 is returned from the F5' do
      it 'raises an exception' do
        allow(IcontrolRest::Client).to receive(:get).and_return bad_response
        expect { api.get_sys_dns }.to raise_error StandardError
      end
    end
  end

  describe '#get' do
    context 'when anything but a 200 is returned from the F5' do
      it 'raises an exception' do
        allow(IcontrolRest::Client).to receive(:get).and_return bad_response
        expect { api.get('/mgmt/tm/sys/dns') }.to raise_error StandardError
      end
    end
  end

  describe '#post' do
    context 'when anything but a 200 is returned from the F5' do
      it 'raises an exception' do
        allow(IcontrolRest::Client).to receive(:post).and_return bad_response
        expect { api.post('/mgmt/tm/sys/dns', key: 'thing') }.to raise_error StandardError
      end
    end
  end
end


spec/spec_helper.rb
# frozen_string_literal: true

require 'simplecov'
SimpleCov.start do
  minimum_coverage 91
  add_filter '/vendor/'
  add_filter '/spec/'
  add_filter '/diagnostics/'
end

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'icontrol_rest'


spec/icontrol_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe IcontrolRest do
  it 'has a version number' do
    expect(IcontrolRest::VERSION).not_to be nil
  end
end


lib/icontrol_rest/utils/logging.rb
# frozen_string_literal: true

require 'logger'

module IcontrolRest
  module Utils
    # Module to handle all logging across this gem
    module Logging
      def self.included(klass)
        klass.extend ClassMethods
        klass.include InstanceMethods
      end

      def self.logger
        @logger ||= NullLogger.new
        @logger.progname = 'IControlRest'
        @logger
      end

      def self.use_logger(logger)
        @logger = logger
      end

      # Module for instance methods to be included when IcontrolRest::Utils::Logging is included by a class.
      module InstanceMethods
        def logger
          IcontrolRest::Utils::Logging.logger
        end
      end

      # Module for class methods to be extended when IcontrolRest::Utils::Logging is included by a class.
      module ClassMethods
        def logger
          IcontrolRest::Utils::Logging.logger
        end
      end

      # A stand-in Logger that does nothing when it is written to.
      #
      # Examples
      #
      #   null_logger = NullLogger.new(STDOUT)
      #   null_logger.info { 'This message will never get writen' }
      #
      # rubocop:disable Lint/MissingSuper
      class NullLogger < Logger
        # Intentionally do nothing
        def initialize(*_args); end

        # Intentionally do nothing
        def add(*_args, &_block); end
      end
      # rubocop:enable Lint/MissingSuper
    end
  end
end


lib/icontrol_rest/client.rb
# frozen_string_literal: true

require 'json'

module IcontrolRest
  # A iControl REST Client for retrieving data from F5s.
  class Client
    include Utils::Logging
    include HTTParty
    format :json

    # A client used to interact with a devices iControl API. Turns method calls into REST commands.
    #
    # host        - The ip or hostname of the host we're trying to reach as a string
    # user        - The username of the account we're connecting with as a string.
    # pass        - the pass of the account we're connecting with.
    # verify_cert - Boolean of whether we want to check certificates.
    # timeout     - Integer of seconds we want the timeout to be set to.
    # sleep       - Number of seconds to sleep after an icontrol call.
    #
    # Examples
    #
    #   api = IcontrolRest::Client.new(host: '1.2.3.4', user: 'user', pass: 'pass', verify_cert: false)
    #   # => #<IcontrolRest::Client:0x007fb953ab7750 @options={ ... }>
    #
    #   api.get_sys_dns
    #   # => {"kind"=>"tm:sys:dns:dnsstate",
    #         "selfLink"=>"https://localhost/mgmt/tm/sys/dns?ver=11.5.4",
    #         "description"=>"configured-by-dhcp",
    #         "nameServers"=>["1.2.3.72", "1.2.3.73"],
    #         "search"=>["domain.com"]}
    #
    # rubocop:disable Metrics/ParameterLists
    def initialize(host:, pass:, user:, timeout: 100, verify_cert: true, sleep: 0, tries: 1)
      self.class.base_uri "https://#{host}"
      @sleep = sleep.to_i
      @tries = tries.to_i
      @options = { basic_auth: { username: user, password: pass }, headers: { 'Content-Type' => 'application/json' },
                   verify: verify_cert, timeout: timeout }
      @retry_handler = proc do |exception, try, _elapsed_time, _next_interval|
        logger.error { "#{exception.class}: #{exception.message} - try #{try}" }
        if exception.instance_of?(JSON::ParserError)
          logger.info { "F5 configuration utility not ready to proceed, sleeping '30' seconds." }
          sleep 30
        end
      end
    end
    # rubocop:enable Metrics/ParameterLists

    # Public    - Delete/Get request method.
    #
    # route     - URI to send to server as a string.
    # headers   - Hash of headers to pass with request.
    #
    # Returns   - response from the server as a hash.
    #
    # Examples
    #
    #   api.get('/mgmt/tm/sys/dns')
    #   # => {"kind"=>"tm:sys:dns:dnsstate",
    #         "selfLink"=>"https://localhost/mgmt/tm/sys/dns?ver=11.5.4",
    #         "description"=>"configured-by-dhcp",
    #         "nameServers"=>["1.2.3.72", "1.2.3.73"],
    #         "search"=>["domain.com"]}
    #
    %w[delete get].each do |action|
      define_method(action.to_s) do |route, headers: {}|
        send_request(action: action, route: route, headers: headers)
      end
    end

    # Public    - Post/Put/Patch request methods.
    #
    # route     - URI to send to server as a string.
    # body      - Hash representing the request body.
    # headers   - Hash of headers to pass with request.
    #
    # Returns   - response from the server as a hash.
    %w[post put patch].each do |action|
      define_method(action.to_s) do |route, body: {}, headers: {}|
        send_request(action: action, route: route, body: body, headers: headers)
      end
    end

    private

    # Private - Handles response and raises error if the response isn't a 200.
    #
    # action  - The HTTP action being executed.
    # route   - The route to send the request to.
    # headers - A Hash representing any headers to be passed with the request.
    # body    - A Hash representing the request body to be passed with the request.
    #
    # Returns - response from the server as a hash.
    #
    # Raises  - RuntimeError when the server returns a response that isn't a 200
    def send_request(action:, route:, headers: {}, body: nil)
      Retriable.retriable on_retry: @retry_handler, tries: @tries do
        body = body.to_json unless body.nil?
        response = self.class.send(action, route, @options.deep_merge(body: body, headers: headers))
        sleep(@sleep) if @sleep.positive?
        return response if response.code == 200

        raise "#{response['code']}: #{response['message']}"
      end
    end

    # Private     - Converts a missing method into a REST request.
    #
    # method_name - Symbol of unknown method called.
    # *args       - An array of arguments passed with the method call.
    #
    # Returns     - Response from the method called.
    def method_missing(method_name, **args)
      super unless respond_to_missing?(method_name)
      route_chain = method_name.to_s.split('_')
      send(route_chain[0].downcase, route(route_chain), **args)
    end

    # Private     - Adds methods prefixed with delete/get/post/put/patch to object.
    #
    # method_name - Symbol of unknown method called.
    #
    # Returns     - true if the object responds to it, or if delete/get..etc.
    def respond_to_missing?(method_name, include_private = false)
      method_name.to_s.start_with?('delete', 'get', 'post', 'put', 'patch') || super
    end

    # Private     - Converts the route chain into a uri.
    #
    # route_chain - An array of route chain.
    #
    # Returns     - A URI as a string.
    def route(route_chain)
      raise 'Empty route chain.' if route_chain.length <= 1

      "/mgmt/tm/#{route_chain[1..route_chain.length].join('/')}"
    end
  end
end
